### DISCUSSION TITLE
Seperate folder for downloaded songs

### DISCUSSION THREAD

**Discussion by Monsterwaill** [KRTirtho/spotube#1700] [Q&A]
### Is there an existing issue for this?

- [ ] I have searched the existing issues

### Pain

Having to go through the data folder to collect your songs is a pisstake imo, having people who want to download a bunch of songs have to root through the data folder trying to get the specific one they downloaded is annoying as fuck

### Suggested solution

My suggestion would be to add a "Downloaded Songs" folder so it is much easier to locate songs that you have downloaded, maybe at a push have new folders created in that folder for different albums.

I would do it myself but I do not know what codebase this application is on nor what application to use to edit the code.

Suggested folder structure:
spotube/downloads/*downloaded song(s)*

if we were to have album folders for the songs you download in that folder, it would go:
spotube/downloads/*downloaded song album(s)*/*downloaded song(s)*

### Self grab

- [ ] I'm ready to work on this issue!

**Comment by KRTirtho** [comment]
Recently, we added support for Library folders. You can change the download folder from "Settinngs" > "Downloads"> "Download Location" to different folder you want to.

Then use the Download folder to only download tracks, and add other folders that contains your already downloaded collection. You can do that using "Library" tab > "Local" > "Add to Library" button

**Comment by SuperGotZa** [comment]
Hi, 

On iOS I am facing this issue where I am unable to download any song. I have tried the stable build and nightly build both. But it does not download, shows an exclamation mark. In the logs, it seems that it does not have access to the root (var) folder. There is option of download folder in settings but when i tap on it, it just opens the file explorer but does not select any folder.
Here’s the log: 

[2025-05-01 10:41:00.714315]---------------------
Bad state: Logs file is empty or non-existent
#0      logsProvider.<anonymous closure> (package:spotube/provider/logs/logs_provider.dart:11)
<asynchronous suspension>
#1      _AsBroadcastStreamController.add (dart:async/broadcast_stream_controller.dart:487)
<asynchronous suspension>

----------------------------------------
[2025-05-01 10:41:08.449077]---------------------
PlatformException(discoveryError, Bonsoir has encountered an error during discovery : -65555: NoAuth (error %s)., -65555, null)

----------------------------------------
[2025-05-01 10:41:09.004525]---------------------
DioException [bad response]: This exception was thrown because the response has a status code of 404 and RequestOptions.validateStatus was configured to throw for this status code.
The status code of 404 has the following meaning: "Client error - the request contains bad syntax or cannot be fulfilled"
Read more about status codes at https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
In order to resolve this exception you typically have either to verify and fix your request code or you have to fix the server code.

#0      DioMixin.fetch (package:dio/src/dio_mixin.dart:520)
<asynchronous suspension>
#1      CustomSpotifyEndpoints.getView (package:spotube/services/custom_spotify_endpoints/spotify_endpoints.dart:80)
<asynchronous suspension>
#2      FutureHandlerProviderElementMixin.handleFuture.<anonymous closure>.<anonymous closure> (package:riverpod/src/async_notifier/base.dart:355)
<asynchronous suspension>

----------------------------------------
[2025-05-01 10:42:07.147657]---------------------
PathAccessException: Creation failed, path = '/var/mobile/Containers/Data/Application/3F50BC95-7EB1-4368-8F04-C37CAAD18D90/Downloads' (OS Error: Operation not permitted, errno = 1)
#0      _checkForErrorResponse (dart:io/common.dart:58)
#1      _Directory.create.<anonymous closure> (dart:io/directory_impl.dart:131)
#2      _rootRunUnary (dart:async/zone.dart:1538)
<asynchronous suspension>
#3      _Directory.create.<anonymous closure>.<anonymous closure> (dart:io/directory_impl.dart:119)
<asynchronous suspension>
#4      DownloadManager.download (package:spotube/services/download_manager/download_manager.dart:81)
<asynchronous suspension>
#5      DownloadManager._startExecution (package:spotube/services/download_manager/download_manager.dart:402)
<asynchronous suspension>

----------------------------------------
[2025-05-01 10:42:20.350897]---------------------
type 'Null' is not a subtype of type 'num' in type cast
#0      _$$InvidiousMonitorImplFromJson (package:invidious/src/models/model.g.dart:471)
#1      new _$InvidiousMonitorImpl.fromJson (package:invidious/src/models/model.freezed.dart:5728)
#2      _$InvidiousMonitorFromJson (package:invidious/src/models/model.freezed.dart:5309)
#3      new InvidiousMonitor.fromJson (package:invidious/src/models/responses/instance.dart:128)
#4      _$$InvidiousInstanceDetailsImplFromJson (package:invidious/src/models/model.g.dart:361)
#5      new _$InvidiousInstanceDetailsImpl.fromJson (package:invidious/src/models/model.freezed.dart:4092)
#6      _$InvidiousInstanceDetailsFromJson (package:invidious/src/models/model.freezed.dart:3858)
#7      new InvidiousInstanceDetails.fromJson (package:invidious/src/models/responses/instance.dart:28)
#8      _$$InvidiousInstanceResponseImplFromJson (package:invidious/src/models/model.g.dart:336)
#9      new _$InvidiousInstanceResponseImpl.fromJson (package:invidious/src/models/model.freezed.dart:3792)
#10     _$InvidiousInstanceResponseFromJson (package:invidious/src/models/model.freezed.dart:3668)
#11     new InvidiousInstanceResponse.fromJson (package:invidious/src/models/responses/instance.dart:11)
#12     InstancesEndpoint.instances.<anonymous closure> (package:invidious/src/endpoints/instances.dart:13)
#13     MappedIterable.elementAt (dart:_internal/iterable.dart:402)
#14     ListIterator.moveNext (dart:_internal/iterable.dart:371)
#15     new _GrowableList._ofEfficientLengthIterable (dart:core-patch/growable_array.dart:190)
#16     new _GrowableList.of (dart:core-patch/growable_array.dart:150)
#17     new List.of (dart:core-patch/array_patch.dart:40)
#18     ListIterable.toList (dart:_internal/iterable.dart:224)
#19     InstancesEndpoint.instances (package:invidious/src/endpoints/instances.dart:17)
<asynchronous suspension>
#20     invidiousInstancesProvider.<anonymous closure> (package:spotube/provider/audio_player/sources/invidious_instances_provider.dart:7)
<asynchronous suspension>
#21     FutureHandlerProviderElementMixin.handleFuture.<anonymous closure>.<anonymous closure> (package:riverpod/src/async_notifier/base.dart:355)
<asynchronous suspension>

----------------------------------------

![IMG_1572](https://github.com/user-attachments/assets/39dd0dc4-dc3f-4823-900b-aa2cdceb67ff)
Added the picture of download error.

https://github.com/user-attachments/assets/5204b415-a829-4126-84d7-a3e8eae3f822
Added the video of me trying to change the download folder.

please let me know if i am doing anything wrong, or if there’s a solution to it. Thanks in advance!

PS: My phone is not jailbroken, I am using eSing to sign ipa files.

**Reply by SuperGotZa** [reply]
Hello @KRTirtho any help on this please?
